### ISSUE TITLE
Introduce back pressure in LogStorageAppender to reduce load into atomix

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3306]
**Is your feature request related to a problem? Please describe.**

As we saw in https://github.com/zeebe-io/zeebe/issues/3257 or https://github.com/zeebe-io/zeebe/issues/3245 if we send high load into atomix either due to internal generated events (https://github.com/zeebe-io/zeebe/issues/3257) or due to lot of requests the raft cluster get disruptive. This means due to the high load atomix is not able to send heartbeats in time and fail over happens. These fails overs can happen multiple times, this or fail over in general can cause other problems, like https://github.com/zeebe-io/zeebe/issues/3274 , https://github.com/zeebe-io/zeebe/issues/3256 or https://github.com/zeebe-io/zeebe/issues/3181. To avoid fail over in the first place we need to reduce the load we send into atomix, since it can't handle it probably.

We tested different back pressure strategies in the log storage append to reduce the load into atomix. With all strategies we saw no fail over, over a long period of time.

In the following I just want to list all tested strategies and different parameters and the results as Grafana Screenshot, were we see the Processing per sec and the related back pressure.

The test configuration is the following:

 * 3 Broker
 * 1 Partition
 * Replication Factor 3
 * 100 wfi/s
   * the workflow  contains 2 Timers (sequentially)
   * they trigger after 5 second this means after 10 seconds 200 timers will trigger regularly
   * This generates a high load on the partition itself

**Comment by Zelldon** [comment]
# Test Results 

We run the last days different benchmarks.

## Simple Byte Limiting

As first we tested a simple approach based on limiting the maximum bytes in flight. This approach has some drawbacks like for larger events, but in general it looks good.

![bytes-based-512-kb](https://user-images.githubusercontent.com/2758593/68081175-fcbe4700-fe09-11e9-8c4e-60b7c9569ba3.png)

## Netflix Concurrency Limiter

After that we tested the netflix concurrency limiter algorithms, which we already use in the CommandAPI. The default settings were not suited to our needed throughput (we got aroud 100 events per sec in the processer, where we need 3k to keep our current throughput)

![vegas-windowed-default](https://user-images.githubusercontent.com/2758593/68081277-bcf85f00-fe0b-11e9-83b5-399ec6ce2f03.png)

That is the reason why we did several benchmarks with permutation the parameters.

### Vegas

We tried the vegas algorithm without the windowed aproach and higher limits, which as first look better. But as you can see after the first back pressure kicks in it was never able to heal from that.

![vegas-non-window-default](https://user-images.githubusercontent.com/2758593/68081243-2d52b080-fe0b-11e9-853b-c4e7d3e715d0.png)

After reading the Javadoc of the Vegas Limiter
```
 * Limiter based on TCP Vegas where the limit increases by alpha if the queue_use is small ({@literal <} alpha)
 * and decreases by alpha if the queue_use is large ({@literal >} beta).
 * 
 * Queue size is calculated using the formula, 
 *  queue_use = limit − BWE×RTTnoLoad = limit × (1 − RTTnoLoad/RTTactual)
 *
 * For traditional TCP Vegas alpha is typically 2-3 and beta is typically 4-6.  To allow for better growth and 
 * stability at higher limits we set alpha=Max(3, 10% of the current limit) and beta=Max(6, 20% of the current limit)
```
We adjusted the alpha and beta functions:
```
    return VegasLimit.newBuilder()
        .alpha(limit -> Math.max(3, limit / 10))
        .beta(limit -> Math.max(3, limit / 5))
```
But unfortunately this does not work in the long run.

![vegas-nw-adjusted-func](https://user-images.githubusercontent.com/2758593/68081244-2deb4700-fe0b-11e9-851a-7a7f09fc0166.png)

Long run:

![vegas-nw-adjusted-func-long](https://user-images.githubusercontent.com/2758593/68081245-2deb4700-fe0b-11e9-90e0-5cf098a4144d.png)


### Gradient

As alternative to Vegas we also tried the Gradient algorithm, which doesn't work either.

![gradient-windowed-l1024](https://user-images.githubusercontent.com/2758593/68081363-23ca4800-fe0d-11e9-8768-1fabf32b90af.png)

### Gradient2

We tried gradient2 as well in the first test it looked much better then the other approaches, that is also the reason why we tried to fine tune the algorithm via the configuration options.

As first try we tested the gradient2 with window limiter and an initial limit of 1024.
![gradient2-first-try](https://user-images.githubusercontent.com/2758593/68081376-5d02b800-fe0d-11e9-93dd-f01fdd2821de.png)

First assumption was that the combination with the window limiter is not ideal and we need a higher queue size. 

So we tested non windowed with queue size of 4:

![gradient2-non-windowed-q4-lw600](https://user-images.githubusercontent.com/2758593/68081384-8de2ed00-fe0d-11e9-8af6-f1b45d06b0d4.png)

**Queue size of 8:**
![gradient2-non-windowed-q8-lw600](https://user-images.githubusercontent.com/2758593/68081385-8de2ed00-fe0d-11e9-8f20-a175aae08b9b.png)


**Queue size of 16:**
![gradient2-non-window-q16-lw600](https://user-images.githubusercontent.com/2758593/68081519-61c86b80-fe0f-11e9-96f2-99b2022ba126.png)


We also tested the queue size (32) and setting the long window (1200 instead of 600) of the gradient2 algorithm, but it seem not to have a real effect. 

![gradient2-non-windowed-q32-lw1200](https://user-images.githubusercontent.com/2758593/68081386-8e7b8380-fe0d-11e9-8133-e96d2be33ff9.png)


**Queue size of 64:**

![gradient2-non-windowed-q64-lw600](https://user-images.githubusercontent.com/2758593/68081387-8e7b8380-fe0d-11e9-9e0c-8ee1cfe47733.png)


**Queue size of 128:**

![gradient2-non-windowed-q128-lw600](https://user-images.githubusercontent.com/2758593/68081388-8e7b8380-fe0d-11e9-8fed-9b09236be4eb.png)

**Queue size of 256**
![gradient2-non-windowed-q256-lw600](https://user-images.githubusercontent.com/2758593/68081391-8e7b8380-fe0d-11e9-92ab-527d2006ae9e.png)

After testing the different queue sizes it seemed either 128 or 256 should be suited.
We then also tried to adjust the RTT tolerance, the default is `1.5`

```
       /**
         * Tolerance for changes in minimum latency.
         * @param rttTolerance Value {@literal >}= 1.0 indicating how much change in minimum latency is acceptable
         *  before reducing the limit.  For example, a value of 2.0 means that a 2x increase in latency is acceptable.
         * @return Chainable builder
         */
```
**Queue size of 128 and RTT 5**
![gradient2-non-windowed-q128-lw600-rtt5](https://user-images.githubusercontent.com/2758593/68081389-8e7b8380-fe0d-11e9-867c-ced4f7ab795b.png)


**Queue size of 128 and RTT 5 long run**

![gradient2-non-windowed-q128-lw600-rtt5-long](https://user-images.githubusercontent.com/2758593/68081390-8e7b8380-fe0d-11e9-9c88-d6ab744c6559.png)


**Queue size of 256 and RTT 5**

![gradient2-non-windowed-q256-lw600-rtt5](https://user-images.githubusercontent.com/2758593/68081392-8f141a00-fe0d-11e9-87dc-5b095e798f4f.png)

Not 100% whether the RTT 5 does have a real effect. As the queue size of 128 looks best for now and we tested everything without a window limiter we wanted to enable the WindowLimiter again.

![gradient2-windowed-q128-lw600-rtt5](https://user-images.githubusercontent.com/2758593/68081490-0d24f080-fe0f-11e9-88ff-7a8b20871889.png)

![gradient2-window-q128-lw600-rtt5](https://user-images.githubusercontent.com/2758593/68081491-0d24f080-fe0f-11e9-9cb0-657370ad3093.png)


## Conclusion 

As we saw in the tests above it seems that the gradient2 algorithm is best suited. I would not go with the vegas nor gradient algorithm as they drop the throughput really hard and vegas doesn't seem to come back. The bytes based approach looks also well, but not sure how problematic it is on larger records. Actually I'm not super happy that we have these up and downs at all but I think this solution is much better then the current state where the cluster gets disruptive.

**Comment by Zelldon** [comment]
After reading a bit about the limit algorithms and reading the Vegas Limiter Code I tried to tune the alpha and beta functions and instead of log10 for increasing and decreasing I used natural log.

```
    return VegasLimit.newBuilder()
        .alpha(limit -> Math.max(3, (int) (limit * 0.5)))
        .beta(limit -> Math.max(6, (int) (limit * 0.9f)))
        .initialLimit(environment.getInt("ZEEBE_INITIAL_APPEND_LIMIT").orElse(1024))
        .maxConcurrency(environment.getInt("ZEEBE_MAX_APPEND_CONCURRENCY").orElse(1024 * 32))
        .increase(limit -> limit + Math.log(limit))
        .decrease(limit -> limit - Math.log(limit))
        .build();
```


With this changes the Vegas results seem much better now.

![vegas-nw-alpha50-beta90-inc-and-dec-log](https://user-images.githubusercontent.com/2758593/68085664-b1c12580-fe43-11e9-9952-cd78c3cece5b.png)

**Comment by Zelldon** [comment]
In the morning I tested the vegas algorithm with different benchmark settings. For example one partition and a high load of 900 create workflow instance per second. Due to the back pressure the Broker was able to withstand the load. It processed 100 workflow instances per second.


**Test 1 partition, 3 Nodes and replica 3:**

![vegas-cluster-1-partition-900wf](https://user-images.githubusercontent.com/2758593/68103959-488bf180-fed9-11e9-9760-07d59bb3db79.png)


I tested also the same load with a larger cluster of 5 Nodes and 8 Partitions.
Here it seems a bit problematic and I saw again some heartbeat misses. It also seem that on these tests the log segment creation take longer then usually (some of them take 7.5 seconds). The leader changes where quite fast and doesn't seem to have a real impact, so on the first look no death partition.

**Test 8 partition, 5 Nodes and replica 3:**

![vegas-cluster-8-partition-90wf-5nodes](https://user-images.githubusercontent.com/2758593/68104201-7faed280-feda-11e9-9cdb-c16c05ab151e.png)

**Test 4 partition, 3 Nodes and replica 3:**

![vegas-cluster-4-partition-900wf-3nodes](https://user-images.githubusercontent.com/2758593/68104200-7f163c00-feda-11e9-95ab-fd63f33fe074.png)
